lib/Sema/CSRanking.cpp

```diff
@@ -1139,17 +1139,6 @@ SolutionCompareResult ConstraintSystem::compareSolutions(
       continue;
     }
 
-    // If one is a function type and the other is a key path, prefer the other.
-    if (type1->is<FunctionType>() != type2->is<FunctionType>()) {
```

Comment:
How do we avoid needing a scoring rule?

Reply:
The change in the implementation of `simplifyKeyPathConstraint` avoids creating a disjunction between the two forms now, so there shouldn't ever be two solutions. It leaves a single `KeyPathConstraint` until there's enough info so that the simplify can pick either function or KeyPath nominal. 

I went this way because the constraint system isn't smart enough / has too much ambiguity to figure out key paths with optional chaining without additional help. Thus the somewhat obscure condition at the end of simplify'ing `!anyComponentsUnresolved || (definitelyKeyPathType && capability == ReadOnly)` -- even if we haven't figured out all the key path component overloads, if we can tell that the literal path will result in a KeyPath nominal and that it'll be read-only (as in an optional chain), then we can immediately bind to the correct nominal type (because no possible component overload can make it non-read-only once a component specifies that it is).

That's essentially a replacement for the logic that used to be in CSGen that immediately bound to a type if there was an optional chain involved instead of creating a `KeyPathConstraint` at all. The constraint system needs that top-down info in order to figure out the bottom-up overloads for the components.

Reply:
What do you get if you write a key path in a context with no type information, like this?

```swift
let kp = \A.property
let _: KeyPath<A, Prop> = kp
```

(There should probably be a test like this if there isn't one already.)

Reply:
If there isn't context that definitively asks for function type vs nominal type, then once the components are all resolved, simplification defaults to the nominal type. (And I'll specifically add tests like that in a moment.)

Reply:
Hmm...could we end up in a situation where we *currently* don't have any context, but the typechecker would discover more context in the future if we left the constraint unsolved? @xedin, any thoughts?